plutus-core/plutus-ir/src/PlutusIR/Purity.hs

```diff
@@ -84,6 +85,9 @@ isPure ver varStrictness = go
             TyAbs {} -> True
             Constant {} -> True
             IWrap _ _ _ t -> go t
+            Let _ _ bs t -> all isPureBinding bs && go t
```

Comment:
Wait, a possibly recursive `Let` can't loop?

Reply:
Good question, I think it can
```
let a (strict) = b
     b (strict) = a
in ...
```

So a group of recursive let-bindings need to be all-nonstrict to be considered pure, right?

Reply:
Great point.

I suggest a dumb approach: just don't consider recursive let bindings to be pure. They're rare anyway. Worth a Note regardless!

Reply:
> just don't consider recursive let bindings to be pure.

That's fine by me. There's also recursive datatype bindings which I guess can also complicates things.

Reply:
I bet they're all at the top level though.